File: web/src/SpeechToText.js
// Copyright 2025 The OpenAgent Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as Setting from "./Setting";
import BrowserSpeechToTextProvider from "./SpeechProvider/BrowserSpeechToTextProvider";
import RemoteSpeechToTextProvider from "./SpeechProvider/RemoteSpeechToTextProvider";
import StreamingSpeechToTextProvider from "./SpeechProvider/StreamingSpeechToTextProvider";
import {bufferToWav} from "./SpeechProvider/AudioUtils";
import i18next from "i18next";

class SpeechToTextHelper {
  constructor(component) {
    this.component = component;
    this.audioChunks = [];
    this.isRecording = false;

    this.browserProvider = new BrowserSpeechToTextProvider(this);
    this.remoteProvider = new RemoteSpeechToTextProvider(this);
    this.streamingProvider = new StreamingSpeechToTextProvider(this);
  }

  initBrowserRecognition(resultCallback, onEndCallback) {
    return this.browserProvider.initBrowserRecognition(resultCallback, onEndCallback);
  }

  // Start an end-to-end streaming cloud STT session. Resolves once the
  // microphone is captured and the websocket has been opened; transcript
  // events arrive on resultCallback over time, and onEndCallback fires
  // when the session ends (server-side completion, user stop, or error).
  startStreaming(store, resultCallback, onEndCallback) {
    return this.streamingProvider.start(store, resultCallback, onEndCallback);
  }

  stopStreaming() {
    this.streamingProvider.stop();
  }

  startRecording() {
    return this.remoteProvider.startRecording();
  }

  stopRecording() {
    return this.remoteProvider.stopRecording();
  }

  stopRecognition() {
    this.browserProvider.stopRecognition();
    this.remoteProvider.stopRecording();
    this.streamingProvider.stop();
  }

  // Convert audio to WAV format
  convertToWav(audioBlob) {
    return new Promise((resolve, reject) => {
      if (!audioBlob || audioBlob.size === 0) {
        reject(new Error("Invalid audio data: empty or null blob"));
        return;
      }

      if (audioBlob.type === "audio/wav") {
        resolve(audioBlob);
        return;
      }

      const AudioContext = window.AudioContext || window.webkitAudioContext;
      const audioContext = new AudioContext({
        sampleRate: 16000, // Force 16kHz sample rate for compatibility
      });

      // Create a file reader to read the blob
      const reader = new FileReader();

      reader.onload = function(e) {
        if (!e.target.result || e.target.result.byteLength === 0) {
          reject(new Error("Audio data is empty after reading file"));
          return;
        }

        audioContext.decodeAudioData(e.target.result)
          .then(buffer => {
            try {
              const wavBuffer = bufferToWav(buffer, 16000); // Force 16kHz sample rate
              const wavBlob = new Blob([wavBuffer], {type: "audio/wav"});
              resolve(wavBlob);
            } catch (wavError) {
              reject(wavError);
            }
          })
          .catch(error => {
            Setting.showMessage("error", i18next.t("general:Failed to decode audio data"));
            reject(error);
          });
      };

      reader.onerror = function(error) {
        reject(error);
      };

      reader.readAsArrayBuffer(audioBlob);
    });
  }

  // Process audio file with speech recognition
  processAudioFile(audioBlob, store, resultCallback) {
    if (!audioBlob || audioBlob.size === 0) {
      Setting.showMessage("error", i18next.t("general:No audio data was captured. Please try again"));
      return;
    }
    this.remoteProvider.processWithCloud(audioBlob, store, resultCallback);
  }

  // Clean up resources
  cleanup() {
    this.browserProvider.cleanup();
    this.remoteProvider.cleanup();
    this.streamingProvider.cleanup();
    this.audioChunks = [];
  }
}

export default SpeechToTextHelper;

File: controllers/speech_to_text.go
// Copyright 2025 The OpenAgent Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//	http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package controllers

import (
	"context"
	"fmt"

	"github.com/the-open-agent/openagent/object"
)

// ProcessSpeechToText
// @Title ProcessSpeechToText
// @Tag STT API
// @Description convert speech to text
// @Param audio formData file true "The audio file to convert to text"
// @Param storeId formData string false "The store ID (one of storeId or providerId is required)"
// @Param providerId formData string false "The provider ID (one of storeId or providerId is required)"
// @Success 200 {object} controllers.SpeechToTextResponse The transcribed text
// @router /process-speech-to-text [post]
func (c *ApiController) ProcessSpeechToText() {
	// Get parameters from form data
	storeId := c.GetString("storeId")
	providerId := c.GetString("providerId")
	if storeId == "" && providerId == "" {
		c.ResponseError(c.T("stt:Missing required parameter: storeId"))
		return
	}

	// Get the audio audioFile from the request
	audioFile, _, err := c.GetFile("audio")
	if err != nil {
		c.ResponseError(fmt.Sprintf("Error getting audio audioFile: %s", err.Error()))
		return
	}
	defer audioFile.Close()

	var provider *object.Provider
	if providerId != "" {
		provider, err = object.GetProvider(providerId)
		if err != nil {
			c.ResponseError(err.Error())
			return
		}
		if provider == nil {
			c.ResponseError(fmt.Sprintf("The provider: %s is not found", providerId))
			return
		}
	} else {
		store, err := object.ResolveStoreFromId(storeId)
		if err != nil {
			c.ResponseError(err.Error())
			return
		}
		if store == nil {
			c.ResponseError(fmt.Sprintf("The store: %s is not found", storeId))
			return
		}

		provider, err = store.GetSpeechToTextProvider()
		if err != nil {
			c.ResponseError(err.Error())
			return
		}
		if provider == nil {
			c.ResponseError(fmt.Sprintf("The speech-to-text provider for store: %s is not found", store.GetId()))
			return
		}
	}

	// Get provider implementation
	providerObj, err := provider.GetSpeechToTextProvider(c.GetAcceptLanguage())
	if err != nil {
		c.ResponseError(err.Error())
		return
	}
	// Process the audio data and get the transcription
	ctx := context.Background()
	text, _, err := providerObj.ProcessAudio(audioFile, ctx, c.GetAcceptLanguage())
	if err != nil {
		c.ResponseError(err.Error())
		return
	}

	// Return the transcribed text
	c.ResponseOk(text)
}
